pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go

```diff
@@ -104,29 +155,142 @@ func hasValidRotateEncryptionKeyStatus(cp *rkev1.RKEControlPlane) bool {
+// update their status when reading from the datastore on startup of the rke2/k3s services.
+func (p *Planner) encryptionKeyRotationRestartLeader(cp *rkev1.RKEControlPlane, clusterPlan *plan.Plan, leader *planEntry) error {
+	status, err := encryptionKeyRotationSecretsEncryptStatusFromPeriodic(leader)
 	if err != nil {
 		return p.enqueueIfErrWaiting(cp, err)
 	}
-	nodePlan := plan.NodePlan{
-		Instructions: []plan.OneTimeInstruction{encryptionKeyRotationRestartInstruction(cp, leader, phase)},
+
+	nodes := collect(clusterPlan, encryptionKeyRotationIsSuitableControlPlane)
+	if len(nodes) == 0 {
+		return p.restartEncryptionKeyRotation(cp) // actually should bail out here, not even a leader
 	}
-	err = assignAndCheckPlan(p.store, fmt.Sprintf("encryption key rotation: [%s]", cp.Status.RotateEncryptionKeysPhase), leader, nodePlan, 0)
-	if err != nil {
-		return p.enqueueIfErrWaiting(cp, err)
+	if len(nodes) == 1 {
+		if nodes[0].Machine.Name != leader.Machine.Name {
+			return p.restartEncryptionKeyRotation(cp) // need to reelect leader
+		}
+		if cp.Status.RotateEncryptionKeysGeneration == cp.Spec.RotateEncryptionKeys.Generation &&
+			status == encryptionKeyRotationReencryptFinished {
+			return p.encryptionKeyRotationFinished(cp) // it is not necessary to restart a leader node after rotation
+		}
+
+		nodePlan := plan.NodePlan{
+			Files: []plan.File{
+				{
+					Content: base64.StdEncoding.EncodeToString([]byte(encryptionKeyRotationRestartControlPlaneScript)),
+					Path:    encryptionKeyRestartApiserverScriptPath(cp),
+				},
+				{
+					Content: base64.StdEncoding.EncodeToString([]byte(encryptionKeyRotationGetAPIServerStatusScript)),
+					Path:    encryptionKeyApiserverStatusScriptPath(cp),
+				},
+			},
+			Instructions:         p.encryptionKeyRotationRestartApiserver(cp),
+			PeriodicInstructions: p.encryptionKeyRotationApiserverStatus(cp),
+		}
+		err = assignAndCheckPlan(p.store, fmt.Sprintf("encryption key rotation: [%s]", cp.Status.RotateEncryptionKeysPhase), leader, nodePlan, 0)
+		if err != nil {
+			return p.enqueueIfErrWaiting(cp, err)
+		}
+
+		if _, ok := cp.Annotations[encryptionKeyApiserverID]; !ok {
+			apiserverID := string(leader.Plan.Output[encryptionKeyRotationRestartApiserverCommand])
+			apiserverID = strings.TrimSuffix(apiserverID, "\n")
+			if len(apiserverID) != 13 { // not a valid container id
+				return p.enqueueAndSkip(cp)
+			}
+			cp.Annotations[encryptionKeyApiserverID] = apiserverID
+			_, err = p.rkeControlPlanes.Update(cp)
+			if err != nil {
+				return err
+			}
+		}
+
+		apiserverStatus := string(leader.Plan.PeriodicOutput[encryptionKeyRotationApiserverStatusCommand].Stdout)
+		apiserverStatus = strings.TrimSuffix(apiserverStatus, "\n")
+		if apiserverStatus == "" {
+			return p.enqueueAndSkip(cp)
+		}
+		if cp.Annotations[encryptionKeyApiserverID] == apiserverStatus {
+			return p.enqueueAndSkip(cp)
+		}
+		if len(apiserverStatus) != 13 { // not a valid container id
```

Comment:
Could you add a longer comment explaining this? Is there a different check we could do that is more intuitive here?

Reply:
I would like to actually, I believe a container ID is always a hex string of 13 characters, but honestly I'd rather use a library but I haven't been able to find one for this purpose. I simply went with the length 13 check because initially I had not set the correct parameters for the command, so it was returning a very big string instead.